src/environment.py

```diff
@@ -0,0 +1,298 @@
+        "MM_SAMLSETTINGS_ENABLE": "true",
+        "MM_SAMLSETTINGS_IDPURL": "https://login.ubuntu.com/saml/",
+        "MM_SAMLSETTINGS_VERIFY": "true",
+        "MM_SAMLSETTINGS_ENCRYPT": "false",  # per POC
+        "MM_SAMLSETTINGS_IDPDESCRIPTORURL": "https://login.ubuntu.com",
+        "MM_SAMLSETTINGS_SERVICEPROVIDERIDENTIFIER": "https://login.ubuntu.com",
+        "MM_SAMLSETTINGS_IDPMETADATAURL": "https://login.ubuntu.com/+saml/metadata",
+        "MM_SAMLSETTINGS_ASSERTIONCONSUMERSERVICEURL": f"https://{site_hostname}/login/sso/saml",
+        "MM_SAMLSETTINGS_LOGINBUTTONTEXT": "Ubuntu SSO",
+        "MM_SAMLSETTINGS_EMAILATTRIBUTE": "email",
+        "MM_SAMLSETTINGS_USERNAMEATTRIBUTE": "username",
+        "MM_SAMLSETTINGS_IDATTRIBUTE": "openid",
+        "MM_SAMLSETTINGS_FIRSTNAMEATTRIBUTE": "fullname",
+        "MM_SAMLSETTINGS_LASTNAMEATTRIBUTE": "",
+        "MM_SAMLSETTINGS_IDPCERTIFICATEFILE": SAML_IDP_CRT,
+        # If not set, we have to install xmlsec1, and Mattermost forks on every login(!).
+        "MM_EXPERIMENTALSETTINGS_USENEWSAMLLIBRARY": use_experimental_saml_library,
+    }
+
+
+def _env_for_smtp(config: dict) -> dict:
+    """Return settings for an outgoing SMTP relay.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        dict of settings for an outgoing SMTP relay.
+    """
+    if any(
+        (
+            setting not in config
+            for setting in (
+                "smtp_host",
+                "smtp_connection_security",
+                "smtp_from_address",
+                "smtp_reply_to_address",
+                "smtp_password",
+                "smtp_user",
+                "smtp_port",
+            )
+        )
+    ):
+        return {}
+
+    if not config["smtp_host"]:
+        return {}
+
+    enable_smtp_auth = "false"
+    if config["smtp_user"] and config["smtp_password"]:
+        enable_smtp_auth = "true"
+
+    # https://github.com/mattermost/mattermost-server/blob/master/model/config.go#L1532
+    return {
+        "MM_EMAILSETTINGS_CONNECTIONSECURITY": config["smtp_connection_security"],
+        "MM_EMAILSETTINGS_ENABLESMTPAUTH": enable_smtp_auth,
+        "MM_EMAILSETTINGS_FEEDBACKEMAIL": config["smtp_from_address"],
+        "MM_EMAILSETTINGS_REPLYTOADDRESS": config["smtp_reply_to_address"],
+        "MM_EMAILSETTINGS_SMTPPASSWORD": config["smtp_password"],
+        "MM_EMAILSETTINGS_SMTPPORT": config["smtp_port"],
+        "MM_EMAILSETTINGS_SMTPSERVER": config["smtp_host"],
+        "MM_EMAILSETTINGS_SMTPUSERNAME": config["smtp_user"],
+    }
+
+
+def missing_config_settings(config: dict) -> list:
+    """Return a list of settings required to satisfy configuration dependencies.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        a list of settings required to satisfy configuration dependencies.
+    """
+    missing = [setting for setting in REQUIRED_SETTINGS if not config.get(setting)]
```

Comment:
What about `[setting for setting in REQUIRED_SETTINGS if setting not in config]`?

Reply:
The [previous code](https://github.com/canonical/mattermost-k8s-operator/blob/main/src/charm.py#L238) was also checking if the value is empty. If I'm doing `a not in b` I'm not checking `bool(b["a"])`